errors: remove ParseError

muzarski opened the pull request:
Ref: https://github.com/scylladb/scylla-rust-driver/issues/519

This PR solves multiple issues:
## No context for request serialization
I introduced a `CqlRequestSerializationError`, with the corresponding variants for each request type. Previously, the request serialization module would return `ParseError`, which did not provide much context.

## Distinguishing between frame ser and deser errors
Decoupled `FrameError` (which originally was overloaded with variants corresponding to serialization and deserialization) into:
- `CqlRequestSerializationError` - in fact, I only moved the variant regarding frame body compression from `FrameError` to `CqlRequestSerializationError`
- `FrameHeaderParseError` - it corresponds to the errors that appear when `Connection::reader()` async task receives a frame, and tries to deserialize its header. It happens before the actual response handler is chosen (based on the stream id).
- `FrameBodyExtensionsParseError` - as the name suggests, it is returned when the driver fails to deserialize body extensions (e.g. frame warnings, or custom payload).

## Test utilities making use of ParseError
There were a couple of test utilities used by `scylla-proxy` that would depend on ParseError. In some cases, if needed, I introduced new error types (e.g. `RequestDeserializationError`). I provided a docstring, noting that this is intended for driver's internal use and testing (it has to be `pub`, because of `scylla-proxy` crate).

## Removing ParseError
With this PR, we could finally remove `ParseError` altogether! I believe that there are still some error-related issues to be addressed, but this is a huge step forward, as this was the most problematic error type.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~[ ] I added relevant tests for new features and bug fixes.~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~[ ] I have adjusted the documentation in `./docs/source/`.~
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
ref: #1056

muzarski commented:
Rebased on https://github.com/scylladb/scylla-rust-driver/pull/1067

wprzytula commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -32,8 +32,10 @@ pub enum FrameDeserializationError {
     ConnectionClosed(usize, usize),
-    #[error(transparent)]
-    StdIoError(#[from] std::io::Error),
+    #[error("Failed to read a frame header: {0}")]
+    HeaderIoError(Arc<std::io::Error>),
```
nit: `a frame header` -> `the frame header` (because there's exactly one header per frame)

wprzytula commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -42,6 +44,16 @@ pub enum FrameDeserializationError {
     Lz4DecompressError(#[from] lz4_flex::block::DecompressError),
 }
 
+impl FrameDeserializationError {
+    pub fn new_header_io_error(err: std::io::Error) -> Self {
+        Self::HeaderIoError(Arc::new(err))
+    }
+
+    pub fn new_body_chunk_io_error(expected_bytes: usize, err: std::io::Error) -> Self {
+        Self::BodyChunkIoError(expected_bytes, Arc::new(err))
+    }
+}
```
Do these have to be `pub`?

muzarski replied:
These are used by `scylla-proxy`, but we could make an exception there and just simply construct corresponding variants directly. I'll do it, since I agree that it does not make much sense for these util functions to be public.

wprzytula commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -66,20 +64,12 @@ pub enum FrameSerializationError {
-    #[error(transparent)]
-    SerializationError(#[from] SerializationError),
+    TryFromIntError(#[from] std::num::TryFromIntError),
 }
```
You add `TryFromIntError` probably because some removed variant contained it. Doesn't this remove the context that the variant provided, leaving us with plain `TryFromIntError` without context?

wprzytula commented on scylla-cql/src/frame/request/mod.rs:
```diff
@@ -74,7 +77,28 @@ pub trait SerializableRequest {
+    fn deserialize(buf: &mut &[u8]) -> Result<Self, RequestDeserializationError>;
+}
+
+/// An error type returned by [`DeserializableRequest::deserialize`].
+/// This is not intented for driver's direct usage. It's a testing utility,
+/// mainly used by `scylla-proxy` crate.
+#[derive(Debug, Error)]
+pub enum RequestDeserializationError {
+    #[error("Low level deser error: {0}")]
+    LowLevelDeserialization(#[from] LowLevelDeserializationError),
+    #[error("Io error: {0}")]
+    IoError(#[from] std::io::Error),
+    #[error("Specified flags are not recognised: {:02x}", flags)]
+    UnknownFlags { flags: u8 },
+    #[error("Named values in frame are currently unsupported")]
+    NamedValuesUnsupported,
+    #[error("Expected SerialConsistency, got regular Consistency: {0}")]
+    ExpectedSerialConsistency(Consistency),
+    #[error(transparent)]
+    BatchTypeParse(#[from] BatchTypeParseError),
+    #[error("Unexpected batch statement kind: {0}")]
+    UnexpectedBatchStatementKind(u8),
 }
```
If it's not intended to be exposed to users, maybe let's make it `doc(hidden)` ?

wprzytula commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -60,9 +60,6 @@ pub enum FrameSerializationError {
-#[derive(Error, Debug)]
-pub enum ParseError {}
-
 #[non_exhaustive]
```
Dieeeee, it's high time.

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -46,30 +46,31 @@
-//! # use scylla_cql::frame::frame_errors::ParseError;
 //! # use scylla_cql::types::deserialize::{DeserializationError, FrameSlice, TypeCheckError};
 //! # use scylla_cql::types::deserialize::value::DeserializeValue;
+//! # use thiserror::Error;
```
Let's not hide this (don't use `#`), because otherwise readers can be confused about the `Error` trait - where it comes from.

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -85,11 +86,18 @@
 //! # use scylla_cql::frame::response::result::ColumnType;
 //! # use scylla_cql::types::deserialize::{DeserializationError, FrameSlice, TypeCheckError};
 //! # use scylla_cql::types::deserialize::value::DeserializeValue;
+//! # use thiserror::Error;
```
ditto

wprzytula commented on scylla-cql/src/types/deserialize/mod.rs:
```diff
@@ -135,32 +137,36 @@
 //! # use scylla_cql::types::deserialize::{DeserializationError, FrameSlice, TypeCheckError};
 //! # use scylla_cql::types::deserialize::value::DeserializeValue;
 //! # use bytes::Bytes;
+//! # use thiserror::Error;
```
ditto

wprzytula reviewed (changes requested):
Mostly good direction, mainly nits.

muzarski commented:
v1.1: fixed a typo in the `RequestDeserializationError` docstring

muzarski commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -40,10 +43,10 @@ pub enum FrameError {
-    QueryParametersSerialization(#[from] QueryParametersSerializationError),
     #[error("Low-level deserialization failed: {0}")]
     LowLevelDeserializationError(#[from] LowLevelDeserializationError),
     #[error("Could not serialize frame: {0}")]
```
> I assume it was introduced in first commit just to make commits atomic?

Exactly. I firstly introduced `QueryParametersSerializationError` - since I changed the return type of `QueryParameters::serialize`, I needed to add it to `ParseError` (returned by `SerializableRequest::serialize`). Then, I introduced two error types that depend on `QPSerializationError`, and thanks to that I could finally remove it from `ParseError`.

muzarski commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -43,7 +43,13 @@ pub enum FrameBodyExtensionsParseError {
+    fn from(value: lz4_flex::block::DecompressError) -> Self {
+        FrameBodyExtensionsParseError::Lz4DecompressError(Arc::new(value))
+    }
 }
```
Good catch

muzarski commented on scylla-cql/src/frame/frame_errors.rs:
```diff
@@ -34,16 +34,16 @@ pub enum FrameError {
+    #[error("Snap decompression error: {0}")]
+    SnapDecompressError(snap::Error),
     #[error("Error decompressing lz4 data {0}")]
     Lz4DecompressError(#[from] lz4_flex::block::DecompressError),
```
What would you suggest then? We could have a common `(De)CompressError` variant for both snap and lz4 - it would not provide any additional context, or maybe it could include a stringified error returned from the corresponding library (and maybe some enum of (de)compression kind - either lz4 or snap).

Lorak-mmk replied:
Why not Arc<dyn Error>?

Lorak-mmk replied:
My phone skipped a part of message - I meant `Arc<dyn Error>`.

muzarski replied:
Yes, I got that. I somehow forgot about this possibility and was trying to come up with some other solution

muzarski commented:
v1.2: addressed @Lorak-mmk 's comments:
- removed `snap::Error` and `lz4_flex::block::DecompressError` types from public API. Replaced them with dynamic `Arc<dyn Error>` in public enum variants.
- added docstring for `FrameBodyExtensionsParseError`

muzarski commented:
@Lorak-mmk I also checked if we use any of these libraries' types in other parts of public API, but didn't find anything. You can double check that, if possible.

muzarski commented:
@Lorak-mmk Rerun the CI, please. A known flaky test `lwt_optimisation::if_lwt_optimisation_mark_offered_then_negotiatied_and_lwt_routed_optimally` failed:
```
`Result::unwrap()` on an `Err` value: DbError(ServerError, "Failed to apply group 0 change due to concurrent modification")
```

muzarski commented:
Can we merge?
